src/storage-client/src/controller/command_wals.rs

```diff
@@ -70,63 +70,109 @@ where
+
+        // Partition metadata into the collections we want to have and those we failed to drop.
+        let (in_use_collections, leaked_collections): (Vec<_>, Vec<_>) =
+            metadata.into_iter().partition(|(id, _, diff)| {
+                assert_eq!(
+                    *diff, 1,
+                    "expected METADATA_COLLECTION to contain reconciled state"
+                );
+
+                self.collection(*id).is_ok()
+            });
+
+        // Get all shard IDs
+        let shard_finalization: BTreeSet<_> = shard_finalization
             .into_iter()
-            .map(
-                |(
-                    id,
-                    // TODO(guswynn): produce the schema for each shard.
-                    super::DurableCollectionMetadata {
-                        remap_shard,
-                        data_shard,
-                    },
-                )| { [(id, [remap_shard, Some(data_shard)])] },
-            )
-            .flatten()
+            .map(|(id, _, diff)| {
+                assert_eq!(
+                    diff, 1,
+                    "expected SHARD_FINALIZATION to contain reconciled state"
+                );
+
+                id
+            })
             .collect();
 
-        // Consider shards in use if they are still attached to a collection.
-        let in_use_shards: BTreeSet<_> = all_shard_data
-            .into_iter()
-            .filter_map(|(id, shards)| self.state.collections.get(&id).map(|_| shards.to_vec()))
-            .flatten()
-            .filter_map(|shard| shard)
+        // Collect all shards from in-use collections
+        let in_use_shards: BTreeSet<_> = in_use_collections
+            .iter()
+            .map(|(_, DurableCollectionMetadata { data_shard, .. }, _)| *data_shard)
             .collect();
 
-        // Determine all shards that were marked to drop but are still in use by
-        // some present collection.
-        let shard_ids_to_keep = registered_shards
+        // Determine if there are any shards that belong to in-use collections
+        // that we have marked for finalization. This is usually inconceivable,
+        // but could happen if we e.g. crash during a migration.
+        let in_use_shards_registered_for_finalization: BTreeSet<_> = shard_finalization
             .intersection(&in_use_shards)
             .cloned()
             .collect();
 
-        // Remove any shards that are currently in use from shard finalization register.
-        self.clear_from_shard_finalization_register(shard_ids_to_keep)
-            .await;
+        // Fixup shard finalization WAL if necessary.
+        if !in_use_shards_registered_for_finalization.is_empty() {
+            self.clear_from_shard_finalization_register(in_use_shards_registered_for_finalization)
+                .await;
+        }
+
+        // If we know about collections that the adapter has forgotten  about,
+        // clean that up.
+        if !leaked_collections.is_empty() {
```

Comment:
The only semantic change is that we now delete things from `METADATA_COLLECTION`? Why are the other changes needed? Is that only to speed it up a bit?

Reply:
Yes, the rest of the changes are more or less just to facilitate parallelizing the scans of the two stash collections.